### PR TITLE
Adds construction gel, for rapid sealing of breaches

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Engineering staff can now use a new substance called construction gel to quickly and rapidly seal breaches.
/:cl:

[Gfycat demo](https://gfycat.com/PopularCourageousFlickertailsquirrel)

Construction gel is a new substance used to rapidly seal breaches, and to create platforms in space that can easily be built on. It's used with a construction gel apparatus, three of which are available from the EngiVend, and one is available from every YouTool when a coin is inserted. They use preloaded canisters, which are also available from the EngiVend and YouTool; the EngiVend has 10, and the YouTool has one double-capacity canister that requires a coin.

The apparatus can select up to 49 tiles worth of space tiles in a rectangular shape by clicking one corner of a rectangle and clicking the other; all tiles between the points will be filled in automatically, and you can either click inside of the area to confirm or outside of it to clear. This will apply construction gel to every tile in the area.

Construction gel's purpose is to serve as a way to quickly fix breaches. A space tile with gel on it will block all atmospheric flow on that tile, and slowly decay over about five minutes. This can be attacked with a floor tile to create plating on top of that gel.

You can also reapply construction gel to existing gel to create a gel wall (and vice-versa). Gel walls are impassable, and can't use floor tiles to create plating. Instead, they can be struck with two metal to create plating _along with_ an anchored girder on top of it.

As a point of note, construction gel does _not_ protect the tile it's on from space; just prevents air from passing through it. Walking through a tile with gel on it is the same as walking through a normal space tile.

### CODE REVIEW

**Comment by WJohn** [comment]
This seems kind of redundant given how quick breaches are to repair, how we have RCDs, and how we have atmos resin.

**Comment by Xhuis** [comment]
Construction gel is very fast, and more widely-available than atmos.
This code was also commissioned, so...

**Comment by calzilla1** [comment]
Kinda like prey's GOO gun?

**Comment by Xhuis** [comment]
Exactly like GLOO, actually. That was the working name for it.

**Comment by Qustinnus** [comment]
Does it work for walls?

**Comment by optimumtact** [comment]
We literally have metal foam for this purpose?

**Comment by Qustinnus** [comment]
They're way dumber, so to speak, as they make walls everywhere and this makes floors.

FTL had smart metal foam grenades that would make walls and floors based on area so if this gel doesn't make walls you should look at that if you want.

**Comment by Ratvar** [comment]
Add malf module for 100pt with permanent construction gels, relise VAL

**Comment by Xhuis** [comment]
@Qustinnus Construction gel only works on space tiles. It can be swapped between gel floors and walls by reapplying gel.

@optimumtact Construction gel is, tl;dr, "I'm you but stronger" in regards to metal foam. It does what metal foam does but is a bit rarer and makes repairing the breach much faster.

**Comment by optimumtact** [comment]
so you've basically made foam redundant?

you may as well remove it then

**Comment by WJohn** [comment]
We have so many layers of redundancy and so many identical options to repair walls and floors.

**Comment by MrPerson** [comment]
I don't see how this is different from just using an RCD

**Comment by Xhuis** [comment]
@WJohn I wasn't actually aware of the existence of atmos resin when I started this. Because they fulfill the same purpose but construction gel is just faster as a whole, I'll probably end up replacing resin with this.

@MrPerson RCDs have fairly limited cartridges and serve a wider role and so can only repair one tile at a time. Gel is for dealing with the immediate aftermath of a hull breach quickly and allowing the actual hole to be fixed in a fast manner; the RCD would ideally be used in conjunction with it, by patching up the actual equipment once the hole is fixed.

**Comment by RemieRichards** [comment]
Pointless, Happens too quickly (all those tiles filled in instantly), Lasts too long (5 minutes?!?!?), Appears to trivialise space travel.

**Comment by jughu** [comment]
what happens if the gloo comes in contact with your hair?

**Comment by CPTANT** [comment]
Why would I ever use metal foam after this is implemented?

**Comment by nicbn** [comment]
Really op

**Comment by Ergovisavi** [comment]
Resin and this do not serve the exact same purpose, given that resin scrubs the air and returns it to a safe temperature,  is transparent, is explicitly in an Atmosian restricted locker, etc. You use resin to both seal off breaches but also to fill a room in to restore atmos, this is just for construction.

That said, I question the need for this. Atmos now has a more unique tool for these situations, and  Atmos should be working with engineers to triage hull breaches  (Which is why I gave them the holo airlocks, to temporarily stop air from leaking out of the staton even if they werent carrying their packs at the time).

Also, walking on gel doesnt protect from space wat

**Comment by Xhuis** [comment]
I'm really attached to the concept of this, so I'm going to stubbornly try to address the (valid!) concerns outlined by people with some changes:

* Each tile will take 3 deciseconds to apply gel too, rather than instantaneously. This will be done through `do_after` checks, and means that sealing a full 49 tiles of space will take about 15 seconds.
* Gel will last for two minutes instead of five. Object integrity unchanged.
* Gel can't be applied directly to space tiles, and needs to use two points of solid ground as a base for application.

@CPTANT As an engineer, you would use it if either the gel apparatuses were sold out or (somehow) there wasn't enough gel to go around. With the other change, it will also be faster, but much less useful.

@Ergovisavi Resin seems to fill the purpose of restoring airflow and preventing air from escaping, which makes construction gel less useful. It does still have a reason to exist, however, in that it serves as a sort of hybrid of metal foam and resin; it prevents air from escaping (but doesn't restore it) and seals the breach (but doesn't protect anyone walking into it.) It's more generally useful than either, and makes the other two fill more niche roles (that will nonetheless occur, of course.)

File: code/game/objects/items/devices/construction_gel.dm [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+	take_damage(rand(0.25, 0.33), sound_effect = FALSE) //Takes a very long time to decay on its own
+	update_icon()
+
+/obj/structure/construction_gel/update_icon()
+	..()
+	cut_overlays()
+	var/glob_size = 1
+	switch(obj_integrity / max_integrity)
+		if(0 to 0.33)
+			glob_size = 1
```
Comment (Cyberboss) [inline code/game/objects/items/devices/construction_gel.dm]: As opposed to setting `1` twice, you could make this the else case

File: code/game/objects/items/devices/construction_gel.dm [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+/obj/structure/construction_gel/wall/attackby(obj/item/I, mob/living/user, params)
+	if(istype(I, /obj/item/stack/sheet/metal))
+		var/obj/item/stack/sheet/metal/M = I
+		if(M.amount < 2)
+			to_chat(user, "<span class='notice'>You need at least two metal to build plating and girders on [src]!</span>")
+			return
+		user.visible_message("<span class='notice'>[user] starts building girders on [src]...</span>", \
+		"<span class='notice'>You begin building girders on [src]...</span>")
+		if(!do_after(user, 30, target = src) || QDELETED(M) || M.amount < 2)
```
Comment (Cyberboss) [inline code/game/objects/items/devices/construction_gel.dm]: use the extra_checks parameter of `do_after` for sanity

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Anonmare** [comment]
I was hoping someone would give this ability to metal foam grenades, but I had no idea on where to begin

**Comment by ike709** [comment]
This is literally just smart metal foam but more complicated.

https://github.com/FTL13/FTL13/pull/451

**Comment by Xhuis** [comment]
I didn't even know that existed. It's probably a better solution; I may decide to port that instead (in which case I'd close this PR.)

**Comment by Fox-McCloud** [comment]
I dunno why there's so many claims of "easy mode" "too easy".

It's already VASTLY easier to demolish and completely trash the station than it is to repair it.

A max cap (or two) and the shuttle usually just gets called as opposed to the crew actually repairing and dealing with the issue.

**Comment by Xhuis** [comment]
Closing this in favor of #29963!
## The concept/thingy/what-have-you is still staying, just in that PR.

**Pull request closed without merging**